### PR TITLE
Travis deploy

### PR DESCRIPTION
**Pull request by werebus**
Use Travis to deploy to Github releases.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+language: node_js
+node_js: node
+script: yarn run compress
+deploy:
+  provider: releases
+  api_key:
+    secure: TQuDynLxEXAfBJeC/dXpXhFzRt8a0DiYFQlfRoKnrAF5P7ZZg0MikEuvjXbwrKA2ArzXVBiFGYEnficjH9zv60ZDHddjDguB4wl72uHqR1lXhgMRzU3s4FTFdFv4g5H7iOOhrqYRh9Sxvgdvf260YxWAMKi8ip6QS1StTi0AEI1d56J+/95aDZZW8aWK9NFuQgZvcm1kBggIvewFKPrDIBEQEWSLeYGu2oT+hmv61s+t+RgUNXhjQ00VuWeHkGHMDzl4ljlJVlC1ZtrnqW9fiYs5VUeJZTX3Ob4e+7rd0noQTyGjaOReLrzuUWLr8MF2fZln9yuiZlhgkvLTmiEdCUgn3BcCFVfri3MgN9bBfuhNclSxVK6qluTPk4eHvPALZSZMjW+1tpPb1TIHa9ibYOfBX70mxSqlBCMjmTd2BGfkDVYFK6FdBLE1Ikc5Z7VQmYPFx8MBPdZBs95eJN6p8CqMhbM+bYHXFIqagXzWCxdps7in0xMUYuWkGjTc2QOA8KAjhorgyNomPY9cvdzDeZhKwFhSvJZQnPxQ3m4x2KMRGskJphTBA6VNwS9GktBq8F5KEQa59gewOcklRhf8kUzNy1QawU8YNrU28wUp8U9hb87NxQzJCJwQZBgMebg+j9EVoRh54jAZ4qNy6fHfBCm8w1XQQbLRLJBP+qZeQRY=
```
Comment (Anbranin) [inline .travis.yml]: So it's cool if the API key is public? I don't know about these things, it just "feels" odd to me.

File: src/index.html [hunk 2]
```diff
@@ -7,15 +7,12 @@
     <link rel="stylesheet" href="https://js.arcgis.com/4.10/esri/css/main.css">
     <link rel="stylesheet" href="index.css">
     <script>
-      // Current path minus everything after the last forward slash
-      var locationPath = location.pathname.replace(/\/[^\/]+$/, "");
```
Comment (Anbranin) [inline src/index.html]: So you don't have to do this because you moved the assets to `src`?

**Comment by werebus** [comment]
Ahem...

File: .travis.yml (hunk 1 above)
Comment (werebus) [reply .travis.yml]: https://docs.travis-ci.com/user/encryption-keys/

Essentially, this is the encrypted value of the API Key - encrypted with a private key that only Travis has.

File: src/index.html (hunk 2 above)
Comment (werebus) [reply src/index.html]: It was an intermediate variable whose use was removed below.

File: .travis.yml (hunk 1 above)
Comment (werebus) [reply .travis.yml]: The value, once decrypted, contains an API key that belongs to @umts-buildbot-gismap. It grants write access to public repositories which that user has write access to. And, that user only has write access to this repository.

**Comment by werebus** [comment]
So, the primary change here is organizational:

* `app/` contains the Typescript that will be compiled with `tsc`
* `src/` contains static files that are copied over verbatim
* `out/` contains the "built" site

I also removed the requirements to have anything "globally" installed in Node and added some helper scripts:  `yarn run build`, `yarn run compress`, `yarn run serve`

Travis only _actually_ does the release upload on `master`, so that part is still a little un-tested. More tweaking might be required.

File: README.md [hunk 3]
```diff
@@ -1,24 +1,34 @@
...
 ```
 
 Compile the application with:
+
 ```
-./compile.sh
+$ yarn run build
+```
+
+The built site will be in the `out/` directory. Serve it with:
```
Comment (LiamBrandt) [inline README.md]: What do you mean by "serve it"?

Comment (werebus) [reply README.md]: Start a webserver? I can be a bit more explicit, I guess. Although, I think the output is (moderately) explanatory:

```
$yarn run serve
yarn run v1.16.0
$ yarn run static out/
$ /home/matt/projects/GISMap/node_modules/.bin/static out/
serving "out/" at http://127.0.0.1:8080
```

**Comment by werebus** [comment]
In my mind, It's about being able to keep separate the build artifact and its sources. If you start out with all the static assets in the `out/` directory and then build the app, you're left with a directory in which _some_ things can be edited by the developer, and _some_ things shouldn't be.  This way, it's easy to say, "don't point your editor at `out/`, it gets destroyed every time you compile the app."

**Comment by werebus** [comment]
I took `src/` as a convention from a few static site generators I've used. But, I'm not married to it.

**Comment by werebus** [comment]
Also, another advantage of keeping `src/` (or, whatever) and `out/` separate is it makes keeping the compiled js from being committed (in the `.gitignore`) a bit simpler.

Other options:
*  Only gitignore .js and .js.map files, but that precludes adding plain-old js files in (for whatever reason)
*  Explicitly un-ignore `index.html`, etc. Although that adds overhead to adding e.g. a favicon or whatever.

File: src/index.html [hunk 4]
```diff
@@ -7,15 +7,12 @@
...
-      // Dojo wants a very explicit path in order to load our JS package
       window.dojoConfig = {
         isDebug: true,
         packages: [
           {
             name: "app",
-            location: location.protocol + "//" + location.host + locationPath + "/app"
+            location: location.origin + "/app"
```
Comment (werebus) [inline src/index.html]: Just realized that this is bad. Won't work if this app is deployed to e.g. a subdirectory.

**Comment by LiamBrandt** [comment]
That makes more sense now, especially the separation of `src/` and `out/`.

I still think `static/` would be a better name for `src/`. `src/` is something you compile, which here is not the case. But I also wouldn't mind if we keep it `src/` so long as it's documented in the readme what that folder is and how `yarn run build` uses it.

File: src/index.html (hunk 2 above)
Comment (werebus) [reply src/index.html]: And also, I was wrong.

**Comment by UmassGIS** [comment]
Just a suggestion to consider: some project use “dist” folder to indicate that the content is ready for distribution and to distinguish from the source/config/dev files.

**Comment by werebus** [comment]
I'm sold on `static/`, but I think I'm going to leave it `out/` rather than `dist/` which, I agree, _is_ a pretty common convention. However, The output directory is arguably an intermediate step in deploying — the final step being a zip file — and also used during development.

**Comment by werebus** [comment]
Alright, I'm going to try merging and see if it actually builds a release.

**Comment by werebus** [comment]
Tada:  https://github.com/umts/GISMap/releases/tag/untagged-c8b6b494b1ddf7aba115

**Pull request merged**